cpu/samd21/cpu.c

```diff
@@ -15,41 +15,151 @@
  *
  * @author      Thomas Eichinger <thomas.eichinger@fu-berlin.de>
  * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Dan Evans <photonthunder@gmail.com>
  * @}
  */
 
 #include "cpu.h"
 #include "periph_conf.h"
+#include "periph_clock_config.h"
 #include "periph/init.h"
 
-#ifndef CLOCK_8MHZ
-#define CLOCK_8MHZ          1
+/**
+ * @name   Default Clock configurations
+ * @{
+ */
+
+/* make sure we have all needed information about the clock configuration */
+#ifndef CLOCK_SRC
+#error "Please provide CLOCK_SRC in your board's perhip_conf.h"
+#endif
+#ifndef CLOCK_CORECLOCK
+#error "Please provide CLOCK_CORECLOCK in your board's periph_conf.h"
+#endif
+
+#ifndef CLOCK_DIV
+#define CLOCK_DIV               (1U)
+#endif
+
+#if CLOCK_SRC == CLOCK_USE_PLL
+#define CLOCK_PLL_MUL           (((CLOCK_CORECLOCK * CLOCK_DIV) / 1000000U) - 1)
+#endif
+
+#if CLOCK_SRC == CLOCK_USE_DFLL
+#define CLOCK_XOSC32K           (32768U)
 #endif
 
-#ifndef GEN2_ULP32K
-#define GEN2_ULP32K         1
+/* If never set then use these defaults */
+#ifndef CLOCK_XOSC32K
+#define CLOCK_XOSC32K           (0)
 #endif
 
 #ifndef VDD
 /**
- * @brief   Set system voltage level in mV (determines flash wait states)
+ * @brief set system voltage level in mV (determines flash wait states)
  *
- * @note    Override this value in your boards periph_conf.h file
- *          if a different system voltage is used.
+ * @note  override this value in your boards periph_conf.h file
+ *        if a different system voltage is used.
  */
-#define VDD                 (3300U)
+#define VDD                     (3300U)
 #endif
 
 /* determine the needed flash wait states based on the system voltage (Vdd)
  * see SAMD21 datasheet Rev A (2017) table 37-40 , page 816 */
 #if (VDD > 2700)
-#define WAITSTATES          ((CLOCK_CORECLOCK - 1) / 24000000)
+#define WAITSTATES              ((CLOCK_CORECLOCK - 1) / 24000000)
 #else
-#define WAITSTATES          ((CLOCK_CORECLOCK - 1) / 14000000)
+#define WAITSTATES              ((CLOCK_CORECLOCK - 1) / 14000000)
 #endif
 
 /**
- * @brief   Configure clock sources and the cpu frequency
+ * @brief setup 1 MHz (8MHz clock/8) clock on Generic Clock Generator 1
+ */
+static bool is_enabled_gen1_1MHz = false;
```

Comment:
I'm unsure if this static variable is really required ? Why do you want to prevent multiple setups ?